### PR TITLE
abci++: Vote extension cleanup

### LINKED ISSUE #9846: Vote extension cleanup
This issue is tracking the cherry-picking of #8402.
The contents of this patch were necessary on preparation for implementing the solution for vote extension propagation (upcoming PR)

### PR DESCRIPTION
**Pull request by sergio-mena**
Closes #9846

This is a PR cherry-picking one single PR: #8402

The contents of this patch were necessary in preparation for implementing the solution for vote extension propagation (which will be in an upcoming PR)

N.B: The changelog will be updated once branch `feature/abci++vef` is merged back into `main`

---

#### PR checklist

- [x] Tests written/updated, or no tests needed
- [x] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [x] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

**Comment by sergio-mena** [comment]
> Seems like a perfect cherry-pick to me 🙂
> 
> Weird that the one test group keeps failing though - seems like it's different tests that fail each time though?

There's `TestMempoolProgressAfterCreateEmptyBlocksInterval` that's driving me crazy, but I think I found a way to reduce the flakines.

I'll push the fix after the dinner (I don't want to push it until I test it well)

File: types/vote.go [hunk 1]
```diff
@@ -224,10 +273,38 @@ func (vote *Vote) ValidateBasic() error {
...
+	if err := vote.ValidateBasic(); err != nil {
+		return err
+	}
+
+	// We should always see vote extension signatures in precommits
+	if vote.Type == tmproto.PrecommitType {
+		// TODO(thane): Remove extension length check once
+		//              https://github.com/tendermint/tendermint/issues/8272 is
+		//              resolved.
+		if len(vote.Extension) > 0 && len(vote.ExtensionSignature) == 0 {
```
Comment (williambanfield) [inline types/vote.go:302]: The details may be a bit foggy to me, but our decision was to ensure that, once vote extensions are enabled, the vote extension must always have a signature regardless of the extension length.

**Comment by williambanfield** [comment]
Looks correct, left a comment making sure my understanding is correct.

File: types/vote.go (hunk 1 above)
Comment (thanethomson) [reply types/vote.go:302]: Looking at #9396, if I understand correctly, that will come in a follow-up PR?

Comment (williambanfield) [reply types/vote.go:302]: Great, then no issue!

**Review by williambanfield — APPROVED** [review]

**Pull request merged**